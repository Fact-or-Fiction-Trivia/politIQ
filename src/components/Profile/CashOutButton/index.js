import React, { useState, useEffect } from 'react';
import moment from 'moment';

import Button from '@material-ui/core/Button';
import Modal from '@material-ui/core/Modal';
import Paper from '@material-ui/core/Paper';
import Close from '@material-ui/icons/Close';

import { db } from '../../../firebase';

const CashOutButton = (props) => {
    const [modalOpen, setModalOpen] = useState(false)
    const [cashoutRequested, setCashoutRequested] = useState(false)

    const toggleModal = () => {
        setModalOpen(!modalOpen);
    }
    
    const requestCashOut = () => {
        // send object with uid, email address, moneywon, and date of cash out request
        // update user information object to say that a cashout has been requested
        const { email, moneyWon, displayName } = props.userInfo
        const date = moment().format('YYYY-MM-DD')
        const uid = props.uid
        db.requestCashOut(uid, date, email, displayName, moneyWon)
            
        const localStorageUser = JSON.parse(localStorage.getItem('authUser'));
        localStorageUser.cashoutRequested = true;
        localStorage.setItem('authUser', JSON.stringify(localStorageUser));
            
        setCashoutRequested(true);
        setModalOpen(false)
      }

      useEffect(() => {
        if (props.userInfo.cashoutRequested !== null) {
            // user cannot cash out if they have already cashed out or if they have not won any money
            const { cashoutRequested, moneyWon } = props.userInfo;
            let requested;
            if (cashoutRequested === "true" || moneyWon === 0) {
              requested = true;
              setCashoutRequested(requested)
            }
        }
      }, [])

    return (
        <>
    {props.userInfo.moneyWon === 0 ? <p id="cashout-instructions">You haven't been awarded any cash prizes yet! Keep playing for your chance to top the leaderboards!</p> : null}
    <Button color="primary" variant="contained" id="cashOut" onClick={toggleModal} disabled={cashoutRequested}>{cashoutRequested && props.userInfo.moneyWon !== 0 ? 'Cash Out Requested' : 'Cash Out' }</Button>
    <Modal
      aria-labelledby="Cash Out"
      aria-describedby="Coming soon - cash out and receive your earnings!"
      open={modalOpen}
      onClose={toggleModal}
      className="cashout-modal"
    >
      <Paper>
        <Close style={{ float: 'right', padding: '1vh', display: 'block'}} onClick={toggleModal}/>
        <h3>Coming soon!</h3>
        <p>You have ${props.userInfo.moneyWon} available.</p>
        <Button variant="contained" color="primary" style={{ marginBottom: '5vh' }} onClick={requestCashOut} id="cashout-confirm">Click Here to Cash Out</Button>
      </Paper>
    </Modal>
    </>
    )
}

export default CashOutButton;