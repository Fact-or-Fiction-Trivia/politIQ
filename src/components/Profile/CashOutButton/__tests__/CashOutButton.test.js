import React from 'react';
import { mount } from 'enzyme';

import CashOutButton from '../';

const uid = "GvM7fEkYp4fcaSuScwixV0xinfT2"
const fakeUserInfo = {
    affiliation: "Republican",
    bio: "Forever thinking about chicken nuggs ",
    cashoutRequested: false,
    displayName: "Hannah",
    email: "hannah.werman@gmail.com",
    emailVerified: false,
    invisibleScore: false,
    isAdmin: false,
    lastActive: "2019-08-06",
    lifetimeEarnings: 5,
    moneyWon: 5,
    soundsOn: true,
    uid: "GvM7fEkYp4fcaSuScwixV0xinfT2"
}

const fakeUserInfo2 = {
    affiliation: "Republican",
    bio: "Forever thinking about chicken nuggs ",
    cashoutRequested: true,
    displayName: "Hannah",
    email: "hannah.werman@gmail.com",
    emailVerified: false,
    invisibleScore: false,
    isAdmin: false,
    lastActive: "2019-08-06",
    lifetimeEarnings: 5,
    moneyWon: 5,
    soundsOn: true,
    uid: "GvM7fEkYp4fcaSuScwixV0xinfT2"
}

describe('<CashOutButton />', () => {
    it('should render correctly and match snapshot', () => {
        const wrapper = mount(<CashOutButton uid={uid} userInfo={fakeUserInfo} />);
        
        expect(wrapper).toMatchSnapshot();

        const button = wrapper.find('button#cashOut')
        expect(button.text()).toEqual('Cash Out')
    })

    it('button click should toggle a modal if the user has not previously requested a cashout and has a nonzero moneyWon', async () => {
        global.localStorage.setItem('authUser', JSON.stringify(fakeUserInfo));

        const wrapper = mount(<CashOutButton uid={uid} userInfo={fakeUserInfo} />);

        const button = wrapper.find('button#cashOut')
        expect(button.props().disabled).toBe(false)
        button.simulate('click')
        wrapper.update()
        const modal = wrapper.find('Modal')

        expect(modal.props().open).toBe(true)

        global.localStorage.clear();
    })

    it('button click should be disabled if user has requested a cashout', async () => {

        const wrapper = mount(<CashOutButton uid={uid} userInfo={fakeUserInfo2} />);

        const button = wrapper.find('button#cashOut');
        // this is probably unnecessary since it seems a bit ridiculous to have to manually update the tests but
        // still in the process of switching over to testing react functional components
        button.props().disabled = wrapper.props().userInfo.cashoutRequested

        expect(button.props().disabled).toBe(true)

        const modal = wrapper.find('Modal')
        modal.props().open = !wrapper.props().userInfo.cashoutRequested

        expect(modal.props().open).toBe(false)
    })
})