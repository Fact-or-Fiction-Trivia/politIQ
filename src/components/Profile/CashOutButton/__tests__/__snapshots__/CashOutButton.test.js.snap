// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CashOutButton /> should render correctly and match snapshot 1`] = `
<CashOutButton
  uid="GvM7fEkYp4fcaSuScwixV0xinfT2"
  userInfo={
    Object {
      "affiliation": "Republican",
      "bio": "Forever thinking about chicken nuggs ",
      "cashoutRequested": false,
      "displayName": "Hannah",
      "email": "hannah.werman@gmail.com",
      "emailVerified": false,
      "invisibleScore": false,
      "isAdmin": false,
      "lastActive": "2019-08-06",
      "lifetimeEarnings": 5,
      "moneyWon": 5,
      "soundsOn": true,
      "uid": "GvM7fEkYp4fcaSuScwixV0xinfT2",
    }
  }
>
  <WithStyles(Button)
    color="primary"
    disabled={false}
    id="cashOut"
    onClick={[Function]}
    variant="contained"
  >
    <Button
      classes={
        Object {
          "colorInherit": "MuiButton-colorInherit-22",
          "contained": "MuiButton-contained-12",
          "containedPrimary": "MuiButton-containedPrimary-13",
          "containedSecondary": "MuiButton-containedSecondary-14",
          "disabled": "MuiButton-disabled-21",
          "extendedFab": "MuiButton-extendedFab-19",
          "fab": "MuiButton-fab-18",
          "flat": "MuiButton-flat-6",
          "flatPrimary": "MuiButton-flatPrimary-7",
          "flatSecondary": "MuiButton-flatSecondary-8",
          "focusVisible": "MuiButton-focusVisible-20",
          "fullWidth": "MuiButton-fullWidth-26",
          "label": "MuiButton-label-2",
          "mini": "MuiButton-mini-23",
          "outlined": "MuiButton-outlined-9",
          "outlinedPrimary": "MuiButton-outlinedPrimary-10",
          "outlinedSecondary": "MuiButton-outlinedSecondary-11",
          "raised": "MuiButton-raised-15",
          "raisedPrimary": "MuiButton-raisedPrimary-16",
          "raisedSecondary": "MuiButton-raisedSecondary-17",
          "root": "MuiButton-root-1",
          "sizeLarge": "MuiButton-sizeLarge-25",
          "sizeSmall": "MuiButton-sizeSmall-24",
          "text": "MuiButton-text-3",
          "textPrimary": "MuiButton-textPrimary-4",
          "textSecondary": "MuiButton-textSecondary-5",
        }
      }
      color="primary"
      component="button"
      disableFocusRipple={false}
      disabled={false}
      fullWidth={false}
      id="cashOut"
      mini={false}
      onClick={[Function]}
      size="medium"
      type="button"
      variant="contained"
    >
      <WithStyles(ButtonBase)
        className="MuiButton-root-1 MuiButton-contained-12 MuiButton-containedPrimary-13 MuiButton-raised-15 MuiButton-raisedPrimary-16"
        component="button"
        disabled={false}
        focusRipple={true}
        focusVisibleClassName="MuiButton-focusVisible-20"
        id="cashOut"
        onClick={[Function]}
        type="button"
      >
        <ButtonBase
          centerRipple={false}
          className="MuiButton-root-1 MuiButton-contained-12 MuiButton-containedPrimary-13 MuiButton-raised-15 MuiButton-raisedPrimary-16"
          classes={
            Object {
              "disabled": "MuiButtonBase-disabled-28",
              "focusVisible": "MuiButtonBase-focusVisible-29",
              "root": "MuiButtonBase-root-27",
            }
          }
          component="button"
          disableRipple={false}
          disableTouchRipple={false}
          disabled={false}
          focusRipple={true}
          focusVisibleClassName="MuiButton-focusVisible-20"
          id="cashOut"
          onClick={[Function]}
          tabIndex="0"
          type="button"
        >
          <button
            className="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-contained-12 MuiButton-containedPrimary-13 MuiButton-raised-15 MuiButton-raisedPrimary-16"
            disabled={false}
            id="cashOut"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-2"
            >
              Cash Out
            </span>
            <NoSsr
              defer={false}
              fallback={null}
            >
              <WithStyles(TouchRipple)
                center={false}
                innerRef={[Function]}
              >
                <TouchRipple
                  center={false}
                  classes={
                    Object {
                      "child": "MuiTouchRipple-child-36",
                      "childLeaving": "MuiTouchRipple-childLeaving-37",
                      "childPulsate": "MuiTouchRipple-childPulsate-38",
                      "ripple": "MuiTouchRipple-ripple-33",
                      "ripplePulsate": "MuiTouchRipple-ripplePulsate-35",
                      "rippleVisible": "MuiTouchRipple-rippleVisible-34",
                      "root": "MuiTouchRipple-root-32",
                    }
                  }
                >
                  <TransitionGroup
                    childFactory={[Function]}
                    className="MuiTouchRipple-root-32"
                    component="span"
                    enter={true}
                    exit={true}
                  >
                    <span
                      className="MuiTouchRipple-root-32"
                    />
                  </TransitionGroup>
                </TouchRipple>
              </WithStyles(TouchRipple)>
            </NoSsr>
          </button>
        </ButtonBase>
      </WithStyles(ButtonBase)>
    </Button>
  </WithStyles(Button)>
  <WithStyles(Modal)
    aria-describedby="Coming soon - cash out and receive your earnings!"
    aria-labelledby="Cash Out"
    className="cashout-modal"
    onClose={[Function]}
    open={false}
  >
    <Modal
      BackdropComponent={[Function]}
      aria-describedby="Coming soon - cash out and receive your earnings!"
      aria-labelledby="Cash Out"
      className="cashout-modal"
      classes={
        Object {
          "hidden": "MuiModal-hidden-31",
          "root": "MuiModal-root-30",
        }
      }
      closeAfterTransition={false}
      disableAutoFocus={false}
      disableBackdropClick={false}
      disableEnforceFocus={false}
      disableEscapeKeyDown={false}
      disablePortal={false}
      disableRestoreFocus={false}
      hideBackdrop={false}
      keepMounted={false}
      manager={
        ModalManager {
          "data": Array [],
          "handleContainerOverflow": true,
          "hideSiblingNodes": true,
          "modals": Array [],
        }
      }
      onClose={[Function]}
      open={false}
    />
  </WithStyles(Modal)>
</CashOutButton>
`;
